### ISSUE TITLE
How does nnU-Net v2 handle the absence of a particular label class during the training and validation process?

### ISSUE THREAD

**Issue by TSbme** [MIC-DKFZ/nnUNet#2438]
Hi,

My data classes are [0,1,2,3,4], and I have split the data into training (1008 cases), validation (218 cases), and testing (124 cases). I used nnU-Net directly for preprocessing, training, validation, and inference, and found that the mean dice score obtained on the validation set was 0.823. 
![image](https://github.com/user-attachments/assets/46ab4630-4afa-495e-ac03-f6bfd2108872)

What does this mean dice score refer to, and how is it calculated? 

Additionally, how does nnU-Net handle the absence of a particular label class during training and validation? Suce as the label classes are [0,2,3]. How does it handle NaN values when calculating the average, and how does it deal with the absence of a class, i.e., NaN values? 

Furthermore, the mean dice score on my test set was 0.827 (calculated by averaging the four categories and then averaging the four average values). I also conducted an external test on 116 cases, where the mean dice score was 0.83. This seems strange to me, and I wonder if there is an issue with my code or how I'm handling NaN values. In the test set, some categories are NaN, and I chose to ignore these images, only calculating the values for non-NaN categories and subtracting from the denominator when averaging at the end (since each category is calculated separately, I didn’t delete the image). 

Here is my code:
```

import os
import SimpleITK as sitk
import medpy.metric.binary as med
import numpy as np
import csv

def evaluate_case(arr: np.ndarray, arr_gt: np.ndarray):
    def compute_metrics(mask_pred, mask_gt):
        if np.any(mask_gt) and np.any(mask_pred):
            dice = med.dc(mask_pred, mask_gt)
            precision = med.precision(mask_pred, mask_gt)
            hd95 = med.hd95(mask_pred, mask_gt)
        else:
            dice, precision, hd95 = np.nan, np.nan, np.nan
        return dice, precision, hd95
    
    mask_gt_rc = (arr_gt == 1).astype(int)
    mask_pred_rc = (arr == 1).astype(int)
    rc_dice, rc_precision, rc_hd95 = compute_metrics(mask_pred_rc, mask_gt_rc)
    
    mask_gt_edema = (arr_gt == 2).astype(int)
    mask_pred_edema = (arr == 2).astype(int)
    edema_dice, edema_precision, edema_hd95 = compute_metrics(mask_pred_edema, mask_gt_edema)

    mask_gt_et = (arr_gt == 3).astype(int)
    mask_pred_et = (arr == 3).astype(int)
    et_dice, et_precision, et_hd95 = compute_metrics(mask_pred_et, mask_gt_et)
    
    mask_gt_net = (arr_gt == 4).astype(int)
    mask_pred_net = (arr == 4).astype(int)
    net_dice, net_precision, net_hd95 = compute_metrics(mask_pred_net, mask_gt_net)

    return rc_dice, rc_precision, rc_hd95, edema_dice, edema_precision, edema_hd95, et_dice, et_precision, et_hd95, net_dice, net_precision, net_hd95

def compute_overall_metrics(total_metrics):
    overall_dice = 0
    overall_precision = 0
    overall_hd95 = 0
    count = 0

    for category in total_metrics:
        if total_metrics[category]['Count'] > 0:
            overall_dice += total_metrics[category]['Dice']
            overall_precision += total_metrics[category]['Precision']
            overall_hd95 += total_metrics[category]['HD95']
            count += total_metrics[category]['Count']
    
    if count > 0:
        return overall_dice / count, overall_precision / count, overall_hd95 / count
    else:
        return np.nan, np.nan, np.nan

if __name__ == '__main__':
    prediction_folder = '/segmentation/DATAFrame/DATASET/nnUNet_raw/Dataset421_BraTS2024/labelsTs'
    ground_truth_folder = '/segmentation/DATAFrame/DATASET/nnUNet_raw/Dataset421_BraTS2024/predict'
    
    total_metrics = {
        'RC': {'Dice': 0, 'Precision': 0, 'HD95': 0, 'Count': 0, 'NaN_Count': 0},
        'Edema': {'Dice': 0, 'Precision': 0, 'HD95': 0, 'Count': 0, 'NaN_Count': 0},
        'ET': {'Dice': 0, 'Precision': 0, 'HD95': 0, 'Count': 0, 'NaN_Count': 0},
        'NET': {'Dice': 0, 'Precision': 0, 'HD95': 0, 'Count': 0, 'NaN_Count': 0},
    }

    with open('/segmentation/inference_internal_test.csv', mode='w', newline='') as file:
        writer = csv.writer(file)
        # Write headers for per-case metrics
        writer.writerow(['File', 'Category', 'Dice', 'Precision', 'HD95', 'Is Dice NaN', 'Is Precision NaN', 'Is HD95 NaN'])

        for pred_file in os.listdir(prediction_folder):
            if pred_file.endswith('.nii.gz'):
                gt_file = pred_file
                pre_path = os.path.join(prediction_folder, pred_file)
                gt_path = os.path.join(ground_truth_folder, gt_file)

                if os.path.exists(gt_path):
                    gt_image = sitk.ReadImage(gt_path)
                    gt_array = sitk.GetArrayFromImage(gt_image)
                    pre_image = sitk.ReadImage(pre_path)
                    pre_array = sitk.GetArrayFromImage(pre_image)
                    
                    rc_dice, rc_precision, rc_hd95, edema_dice, edema_precision, edema_hd95, et_dice, et_precision, et_hd95, net_dice, net_precision, net_hd95 = evaluate_case(pre_array, gt_array)
                    
                    # Update metrics and track NaNs
                    def update_metrics_and_log(category, dice, precision, hd95):
                        if not np.isnan(dice):
                            total_metrics[category]['Dice'] += dice
                            total_metrics[category]['Precision'] += precision
                            total_metrics[category]['HD95'] += hd95
                            total_metrics[category]['Count'] += 1
                        else:
                            total_metrics[category]['NaN_Count'] += 1
                        writer.writerow([pred_file, category, dice, precision, hd95, np.isnan(dice), np.isnan(precision), np.isnan(hd95)])
                    
                    update_metrics_and_log('RC', rc_dice, rc_precision, rc_hd95)
                    update_metrics_and_log('Edema', edema_dice, edema_precision, edema_hd95)
                    update_metrics_and_log('ET', et_dice, et_precision, et_hd95)
                    update_metrics_and_log('NET', net_dice, net_precision, net_hd95)

                    print(f'Processed {pred_file}: RC Dice={rc_dice}, Edema Dice={edema_dice}, ET Dice={et_dice}, NET Dice={net_dice}')
                    print(f'Processed {pred_file}: RC Precision={rc_precision}, Edema Precision={edema_precision}, ET Precision={et_precision}, NET Precision={net_precision}')
                    print(f'Processed {pred_file}: RC HD95={rc_hd95}, Edema HD95={edema_hd95}, ET HD95={et_hd95}, NET HD95={net_hd95}')
                else:
                    print(f'Prediction file {pred_file} does not have a corresponding ground truth file in {ground_truth_folder}')
        
        # Write headers for the average metrics
        writer.writerow(['Category', 'Average Dice', 'Average Precision', 'Average HD95', 'Total NaN Count'])

        for category in ['RC', 'Edema', 'ET', 'NET']:
            if total_metrics[category]['Count'] > 0:
                avg_dice = total_metrics[category]['Dice'] / total_metrics[category]['Count']
                avg_precision = total_metrics[category]['Precision'] / total_metrics[category]['Count']
                avg_hd95 = total_metrics[category]['HD95'] / total_metrics[category]['Count']
            else:
                avg_dice, avg_precision, avg_hd95 = np.nan, np.nan, np.nan

            writer.writerow([category, avg_dice, avg_precision, avg_hd95, total_metrics[category]['NaN_Count']])
            print(f'Average {category} Dice: {avg_dice}, Precision: {avg_precision}, HD95: {avg_hd95}, NaN Count: {total_metrics[category]["NaN_Count"]}')

        overall_dice, overall_precision, overall_hd95 = compute_overall_metrics(total_metrics)
        writer.writerow(['Overall', overall_dice, overall_precision, overall_hd95, sum(total_metrics[cat]['NaN_Count'] for cat in total_metrics)])
        print(f'Overall Dice: {overall_dice}, Precision: {overall_precision}, HD95: {overall_hd95}')
        print('Results have been saved to inference_G.csv')

```

**Comment by aymuos15** [comment]
`Additionally, how does nnU-Net handle the absence of a particular label class during training and validation?`
It is not compulsory to have each label in each case. This is for training/val in general.

`Furthermore, the mean dice score on my test set was 0.827 (calculated by averaging the four categories and then averaging the four average values). I also conducted an external test on 116 cases, where the mean dice score was 0.83. This seems strange to me,`
Why would this be strange?

About the calculation, simply leave out the nans for those cases. you may average X cases for a certain label and Y cases for another.

**Comment by TSbme** [comment]
Thank you so much!!
In the validation process of nnU-Net, if a certain class has a very small target, resulting in poor segmentation performance and the Dice coefficient for that class being NaN, how is this handled in the final averaging? Is NaN set to 0 and included in the total for averaging, or are these values directly ignored, and only the non-NaN values averaged?
